cpp/src/arrow/compute/kernels/boolean.cc

```diff
@@ -43,34 +43,35 @@ namespace compute {
 class InvertKernel : public UnaryKernel {
   Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
     DCHECK_EQ(Datum::ARRAY, input.kind());
+    constexpr int64_t kZeroDestOffset = 0;
 
     const ArrayData& in_data = *input.array();
-    ArrayData* result;
-
-    out->value = ArrayData::Make(boolean(), in_data.length);
-    result = out->array().get();
+    std::shared_ptr<ArrayData> result = out->array();
+    result->type = boolean();
 
-    // Allocate or copy bitmap
+    // Handle validity bitmap
     result->null_count = in_data.null_count;
-    std::shared_ptr<Buffer> validity_bitmap = in_data.buffers[0];
+    const std::shared_ptr<Buffer>& validity_bitmap = in_data.buffers[0];
```

Comment:
`const Buffer&` could even be more idiomatic

Reply:
In this case I need the reference to the shared_ptr because I copy it below (I suppose I could refactor slightly to avoid this).  I changed the one below since it is not needed there.